File: my-folder/0084-largest-rectangle-in-histogram/solution.java

class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int[] pref = new int[n];
        pref[0] = -1;
        for (int i = 1; i < n; i++) {
            int j = i - 1;
            while (j >= 0 && heights[j] >= heights[i]) {
                j = pref[j];
            }

            pref[i] = j;
        }

        int[] suf = new int[n];
        suf[n - 1] = n;
        for (int i = n - 2; i >= 0; i--) {
            int j = i + 1;
            while (j < n && heights[j] >= heights[i]) {
                j = suf[j];
            }

            suf[i] = j;
        }

        int max = 0;
        for (int i = 0; i < n; i++) {
            int area = heights[i] * (suf[i] - pref[i] - 1);
            max = Math.max(max, area);
        }

        return max;
    }
}

File: my-folder/0084-largest-rectangle-in-histogram/solution.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        n = len(heights)
        nextSmall = [(n) for _ in range(n)]
        prevSmall = [-1 for _ in range(n)]

        st = []

        # nextSmall pass
        for i in range(n):
            while len(st) > 0 and heights[st[-1]] > heights[i]:
                idx = st.pop()
                nextSmall[idx] = i 
            
            st.append(i)
        st = []        
        # prevSmall
        for i in range(n-1, -1, -1):
            while len(st) > 0 and heights[st[-1]] > heights[i]:
                idx = st.pop()
                prevSmall[idx] = i
            st.append(i)
        
        ans = -1

        
        dbug = True
        if dbug:
            for i in range(n):
                print(str(nextSmall[i]), end=" ")
            
            print("")
            
            for i in range(n):
                print(prevSmall[i], end =" ")

            print("")

        for i in range(n):
            r = nextSmall[i]
            l = prevSmall[i]

            ans = max(
                ans, 
                heights[i] * (r - l -1)
            )

        return ans
